File: oop/06-inheritance-vs-composition/csharp/Composition/SavingsAccount.cs
/*
 * Demonstrates: Inheritance vs Composition - Composition Approach
 * Concept: SavingsAccount uses an injected IInterestCalculator instead of inheriting interest behaviour
 */

namespace FinanceDomain.InheritanceVsComposition.Composition;

public class SavingsAccount
{
    private decimal _balance;
    private readonly IInterestCalculator _interestCalculator;

    public string AccountNumber { get; }
    public string AccountHolder { get; }
    public decimal MinimumBalance { get; }
    public decimal Balance => _balance;

    public SavingsAccount(
        string accountNumber,
        string accountHolder,
        decimal openingBalance,
        decimal minimumBalance,
        IInterestCalculator interestCalculator)
    {
        AccountNumber = accountNumber;
        AccountHolder = accountHolder;
        _balance = openingBalance;
        MinimumBalance = minimumBalance;
        _interestCalculator = interestCalculator;
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        _balance += amount;
    }

    public void Withdraw(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be greater than zero.");

        decimal resultingBalance = _balance - amount;

        if (resultingBalance < MinimumBalance)
            throw new InvalidOperationException($"Cannot withdraw. Minimum balance of R{MinimumBalance:F2} required.");

        _balance = resultingBalance;
    }

    public decimal CalculateInterest()
    {
        return _interestCalculator.CalculateInterest(_balance);
    }

    public void ApplyInterest()
    {
        decimal interest = CalculateInterest();
        _balance += interest;
    }
}

File: oop/06-inheritance-vs-composition/typescript/composition/savingsAccount.ts
/*
 * Demonstrates: Inheritance vs Composition - Composition Approach
 * Concept: SavingsAccount uses an injected IInterestCalculator instead of inheriting interest behaviour
 */

import { IInterestCalculator } from "./interestCalculator";

export class SavingsAccount {
    private _balance: number;
    private readonly interestCalculator: IInterestCalculator;

    public readonly accountNumber: string;
    public readonly accountHolder: string;
    public readonly minimumBalance: number;

    constructor(
        accountNumber: string,
        accountHolder: string,
        openingBalance: number,
        minimumBalance: number,
        interestCalculator: IInterestCalculator
    ) {
        this.accountNumber = accountNumber;
        this.accountHolder = accountHolder;
        this._balance = openingBalance;
        this.minimumBalance = minimumBalance;
        this.interestCalculator = interestCalculator;
    }

    get balance(): number {
        return this._balance;
    }

    deposit(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }
        this._balance += amount;
    }

    withdraw(amount: number): void {
        if (amount <= 0) {
            throw new Error("Amount must be greater than zero.");
        }

        const resultingBalance = this._balance - amount;

        if (resultingBalance < this.minimumBalance) {
            throw new Error(`Cannot withdraw. Minimum balance of R${this.minimumBalance.toFixed(2)} required.`);
        }

        this._balance = resultingBalance;
    }

    calculateInterest(): number {
        return this.interestCalculator.calculateInterest(this._balance);
    }

    applyInterest(): void {
        const interest = this.calculateInterest();
        this._balance += interest;
    }
}
